jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/GraphOrdering.java

```diff
@@ -201,9 +204,22 @@ public V getVertex(int vertexNumber)
      */
     public E getEdge(int v1Number, int v2Number)
     {
+        
+        if (cacheEdges) {
+            final E cache = edgeCache[v1Number*vertexCount+v2Number];
```

Comment:
In the case where the edge doesn't exist, we will always have to re-confirm that (making the caching ineffective).  If we agree to always combine the edge existence cache with the edge reference cache, then we can make use of the edge existence cache to solve that.

Reply:
I don't think `getEdge` is ever called if `!hasEdge`. That being said, would you want something like:
```
if(cacheEdges){
    if(!hasEdge(v1Number, v2Number) return null; // assuming this is the correct return value?
    final E cache = edgeCache[v1Number*vertexCount+v2Number];
    ....
}
```

Reply:
It's possible to further optimize it to avoid calling both containsEdge and getEdge on the first (uncached) call.  In other words:

1.  check cache bit
2.  if already cached, just do array lookup
3.  if not cached yet, call getEdge, and update both caches based on result (getEdge returns null whenever containsEdge would have returned false)

Reply:
I actually think one may want to do this in `hasEdge` - namely instead of calling `containsEdge`, call `getEdge` and update both cache matrices.

I believe the typical use case is calls to `hasEdge` followed by `getEdge` calls, so I'd like to optimize that path.

I'll implement it once #914 lands

Reply:
That works too.